Gems/EMotionFX/Code/EMotionFX/Source/BlendTreeFabrikNode.cpp

```diff
@@ -0,0 +1,660 @@
+
+        const ActorInstance* actorInstance = m_animGraphInstance->GetActorInstance();
+        const Actor* actor = actorInstance->GetActor();
+        const Skeleton* skeleton = actor->GetSkeleton();
+
+        // don't update the next time again
+        m_nodeIndices.clear();
+        m_alignNodeIndex = InvalidIndex;
+        m_bendDirNodeIndex = InvalidIndex;
+        m_endEffectorNodeIndex = InvalidIndex;
+        SetHasError(true);
+
+        const AZStd::string& rootJointName = fabrikIKNode->GetRootJointName();
+        // Find the end joint.
+        const AZStd::string& endJointName = fabrikIKNode->GetEndJointName();
+        if (rootJointName.empty() || endJointName.empty() || rootJointName == endJointName)
+        {
+            return;
+        }
+        const Node* rootJoint = skeleton->FindNodeByName(rootJointName);
+        if (!rootJoint)
+        {
+            return;
+        }
+
+        for (Node* joint = skeleton->FindNodeByName(endJointName); joint != rootJoint; joint = joint->GetParentNode())
+        {
+            if (joint == nullptr)
+            {
+                // it means the root node is not the ancester of end joint
+                return;
+            }
+            m_nodeIndices.push_back(joint->GetNodeIndex());
+        }
+        m_nodeIndices.push_back(rootJoint->GetNodeIndex());
+        AZStd::reverse(m_nodeIndices.begin(), m_nodeIndices.end());
+
+        // Get the end effector joint.
+        const AZStd::string& endEffectorJointName = fabrikIKNode->GetEndEffectorJointName();
+        const Node* endEffectorJoint = skeleton->FindNodeByName(endEffectorJointName);
+        if (endEffectorJoint)
+        {
+            m_endEffectorNodeIndex = endEffectorJoint->GetNodeIndex();
+        }
+
+        // Find the bend direction joint.
+        const AZStd::string& bendDirJointName = fabrikIKNode->GetBendDirJointName();
+        const Node* bendDirJoint = skeleton->FindNodeByName(bendDirJointName);
+        if (bendDirJoint)
+        {
+            m_bendDirNodeIndex = bendDirJoint->GetNodeIndex();
+        }
+
+        // lookup the actor instance to get the alignment node from
+        const NodeAlignmentData& alignToJointData = fabrikIKNode->GetAlignToJointData();
+        const ActorInstance* alignInstance = m_animGraphInstance->FindActorInstanceFromParentDepth(alignToJointData.second);
+        if (alignInstance)
+        {
+            if (!alignToJointData.first.empty())
+            {
+                const Node* alignJoint = alignInstance->GetActor()->GetSkeleton()->FindNodeByName(alignToJointData.first.c_str());
+                if (alignJoint)
+                {
+                    m_alignNodeIndex = alignJoint->GetNodeIndex();
+                }
+            }
+        }
+
+        SetHasError(false);
+    }
+
+    BlendTreeFabrikNode::BlendTreeFabrikNode()
+        : AnimGraphNode()
+        , m_rotationEnabled(false)
+        , m_relativeBendDir(true)
+        , m_extractBendDir(false)
```

Comment:
I see we initialize them here, but it woudln't be necessary if we did so in the header.